src/cdsetool/credentials.py

```diff
@@ -83,19 +91,37 @@ def __init__(
 
         self.__ensure_tokens()
 
-    def get_session(self):
+    def get_session(self, authorization=True):
         """
-        Returns a session with the credentials set as the Authorization header
+        Returns a configured session. When authorization is true, credentials
+        are also set as the Authorization header
         """
-        self.__ensure_tokens()
+        if authorization:
+            self.__ensure_tokens()
 
         session = requests.Session()
-        session.headers.update({"Authorization": f"Bearer {self.__access_token}"})
+        session.mount("http://", HTTPAdapter(max_retries=self.__retries))
+        session.mount("https://", HTTPAdapter(max_retries=self.__retries))
+        if authorization:
+            session.headers.update({"Authorization": f"Bearer {self.__access_token}"})
         return session
 
     def __token_exchange(self, data):
+        # Make a session that will retry post, respecting the retry-after
+        # header when we get a 503 and a few other temporary failures.
+        session = self.get_session(authorization=False)
+        retries = Retry(
+            total=2,
+            backoff_factor=0.5,
+            allowed_methods=None,
+            raise_on_status=False,
+            status_forcelist=Retry.RETRY_AFTER_STATUS_CODES,
+        )
+        session.mount("http://", HTTPAdapter(max_retries=retries))
+        session.mount("https://", HTTPAdapter(max_retries=retries))
```

Comment:
I'm not sure i see the point in modifying the `get_session` method to be usable here (the `authorization=True` addition on L94) when you end up repeating the bulk of the `get_session` method here. 

This could be slimmed down with no difference in function by reverting the changes made to `get_session` and applying this change
```suggestion
        session = requests.Session()
        retries = Retry(
            total=2,
            backoff_factor=0.5,
            allowed_methods=None,
            raise_on_status=False,
            status_forcelist=Retry.RETRY_AFTER_STATUS_CODES,
        )
        session.mount("http://", HTTPAdapter(max_retries=retries))
        session.mount("https://", HTTPAdapter(max_retries=retries))
```

Reply:
You have a point about code duplication, but line 183 uses `get_session`, and so does `download_feature`, so I don't think those changes should be reverted.

How about passing in an optional Retry object to get_session instead? That way the code you marked can retry a little and for any method, and the rest of CDSETool uses the (presumably) safe defaults from urllib3?

Pushed code is completely untested, I have no idea if I can assign parameter values with `retries=retries or self.__retries`.

Reply:
>  but line 183 uses get_session, and so does download_feature, so I don't think those changes should be reverted.

I can see where you are going with this, but i think the reusable session code should be moved elsewhere. The entire point of the `Credentials` class is to generate sessions with authorization enabled, and the `get_session` method has this implication built in.
Adding the option to disable this more or less bypasses the class's entire function.

I think the `get_session` code as it is right now would fit better in a generic http backend class along the lines of #93 . This also has the additional benefit of providing a convenient way of configuring retries globally, through a mechanism similar to the suggested global `proxies` config. 

I would either go the way of the HTTP class, or provide a private method for generating sessions with retry and using that in the public `get_session` method. This way the public API of the `Credentials` class does not get tainted by private code.

Reply:
I think the HTTP class is a much bigger change, so I have updated this PR to keep the internal use in a private method as you suggest.